File: app/src/main/java/it/unical/mat/embasp/asp/AnswerSet.java
package it.unical.mat.embasp.asp;

import java.lang.reflect.InvocationTargetException;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AnswerSet {

	private List<String> value;
	private Map<Integer,Integer> weight_map;
	private Set<Object> atoms;


	public AnswerSet (List<String> output) {
		value = output;
		weight_map = new HashMap<>();
	}



	public AnswerSet(List<String> value, Map<Integer, Integer> weightMap) {
		this.value = value;
		this.weight_map = weightMap;
	}


	public List<String> getAnswerSet() {
		return Collections.unmodifiableList(value);
	}

	public Set<Object> getAtoms() throws IllegalAccessException, IllegalArgumentException, InvocationTargetException, NoSuchMethodException, SecurityException, InstantiationException{
		if(atoms == null){
			atoms = new HashSet<>();
			ASPMapper mapper = ASPMapper.getInstance();
			for(String atom : value){
				Object obj = mapper.getObject(atom);
				if(obj !=null )
					atoms.add(obj);
			}
		}

		return atoms;
	}

	@Override
	public String toString() {

		return value.toString();
	}
}

File: CSharp version/EmbASP/it/unical/mat/embasp/languages/asp/AnswerSet.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;


namespace it.unical.mat.embasp.languages.asp
{

	public class AnswerSet
	{

		private readonly IList<string> value;
		private readonly IDictionary<int, int> weight_map;
		private ISet<object> atoms;

		public AnswerSet(IList<string> output)
		{
			value = output;
			weight_map = new Dictionary<int, int>();
		}

		public AnswerSet(IList<string> value, IDictionary<int, int> weightMap)
		{
			this.value = value;
			weight_map = weightMap;
		}

		public virtual IList<string> getAnswerSet()
		{
      return new ReadOnlyCollection<string>(value);
		}

		public virtual ISet<object> Atoms
		{
			get
			{
				if (atoms == null)
				{
					atoms = new HashSet<object>();
					ASPMapper mapper = ASPMapper.Instance;
					foreach (String atom in value)
					{
            //Console.WriteLine("ATOM NAME {0}", atom);
						object obj = mapper.getObject(atom);
						if (obj != null)
						{
							atoms.Add(obj);
						}
					}
				}
    
				return atoms;
			}
		}

		public virtual IDictionary<int, int> LevelWeight
		{
			get
			{
				return weight_map;
			}
		}

		public virtual IDictionary<int, int> Weights
		{
			get
			{
        return new ReadOnlyDictionary<int,int>(weight_map);
			}
		}

		public override string ToString()
		{
			return value.ToString();
		}
	}

}